File: js/palindromePartitioning.js
/**
 * 
Given a string s, partition s such that every substring of the partition is a palindrome.

Return all possible palindrome partitioning of s.

Example:

Input: "aab"
Output:
[
  ["aa","b"],
  ["a","a","b"]
]
 */

/**
 * @param {string} s
 * @return {string[][]}
 */
var partition = function(s) {
  if (!s) {
    return [];
  }
  let res = [];
  helper(s, 0, [], res);
  return res;

  function helper(s, idx, currentPartition, res) {
    // idx is the start of the next partition point
    if (idx >= s.length) {
      res.push(currentPartition);
      return;
    }
    for (let i = idx; i < s.length; i++) {
      let currentSubstirng = s.substring(idx, i + 1);
      if (isPalindrome(currentSubstirng)) {
        helper(s, i + 1, [...currentPartition, currentSubstirng], res);
      }
    }
  }
};

const isPalindrome = function(s) {
  return (
    s ===
    s
      .split("")
      .reverse()
      .join("")
  );
};

File: problems/palindromePartitioning.py
"""
Given a string s, partition s such that every substring of the partition is a palindrome.

Return all possible palindrome partitioning of s.

Example:

Input: "aab"
Output:
[
  ["aa","b"],
  ["a","a","b"]
]
"""
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        res = []
        self.helper(0, [], res, s)
        return res

    def helper(self, idx, currRes, res, s):
        if idx == len(s):
            res.append(currRes.copy())
        else :
            for i in range(idx, len(s)):
                currentString = s[idx:i+1]
                if(self.isPalindrome(currentString)):
                    currRes.append(currentString)
                    self.helper(i+1, currRes, res, s)
                    currRes.pop()

    def isPalindrome(self, s):
        return s == s[::-1]